Updated to maintain roughly the same api, but remove monkeypatches

webark opened the pull request:
and with minimum useage of private apis.

Also enables source mapping and hot realoding,
and sets us up for supporting in addon usage.

* Need to add back in the accptance test after removed
  it due to such severe ember upgrade differences.
* Need to add support for plain css useage
* Need to better differentiate between pod-styles.scss, and pod-styles.sass,
  rather then just not accepting sass files and just scss styles instead.
* Need better testing around empty states and precompilers other then sass

webark commented:
This will probably change a bit, but wanted to get this convo started. Would like to add back in some tests, but will hopefully fix #122, #109, #96, #95, #82, and others.

webark commented:
And since I know that you have been concerned with the performance with this addon @stefanpenner, I would appreciate any feedback or suggestions that you have as well.

stefanpenner commented on addon/initializers/component-styles.js:
```diff
@@ -0,0 +1,35 @@
+  });
+
+  Component.reopen({
+    addPodStyleName: Ember.on('init', function() {
```
use `init` + super instead, better performance.

stefanpenner commented on lib/pod-names.js:
```diff
@@ -0,0 +1,32 @@
+PodNames.prototype.constructor = PodNames;
+function PodNames(inputNode, options) {
+  options = options || {};
+  Plugin.call(this, [inputNode], {
```
likely want to revisit this as a persistentOutput filter, so the build output remains stable and doesn't invalidate downstream. This can come as a later addition, but will likely be important to maintain performance.

webark replied:
@stefanpenner a little curious as to how to do this with a filter. I'm wanting to take a set of files, and turn part of their paths into a json blob that can get exported to be imported after build into the initializer. If the `processString` method outputted a key value pair, could you then use the `postProcess` method to join all of those into a single output?

stefanpenner replied:
persistentOutput is a mode one can put a broccoli plugin in, this allows it to take over management of its output directory and what enabled merge-trees + persistent-filter + funnel to get a 10x speed boost.

stefanpenner replied:
relevant pr:

* https://github.com/broccolijs/broccoli-funnel/pull/44
* https://github.com/broccolijs/broccoli-merge-trees/pull/34

stefanpenner replied:
basically in build, you want to calculate the operations required to "update" the persistent output directory.

fs-tree is a module written to help compare the current input dir, with the last, which allows one to write functions like [this one](https://github.com/stefanpenner/broccoli-persistent-filter/blob/master/index.js#L62-L91) which are extremely efficient.

I am sorry the current public API for these types of performant plugins is so shitty, I haven't gotten around to implementing my higher level + more digestable ideas.

webark replied:
OK I'll look into these too.

stefanpenner commented on lib/pod-style.js:
```diff
@@ -0,0 +1,18 @@
+var Filter = require('broccoli-filter');
```
lets use persistent-filter, as it uses persistentOutput and also offers a cache shared between restarts.

topaxi commented:
This looks very good!

To support plain css and fix https://github.com/ebryn/ember-component-css/issues/107 you could use the removed [postcss-selector-namespace](https://github.com/topaxi/postcss-selector-namespace) plugin together with [postcss-scss](https://github.com/postcss/postcss-scss) and [postcss-less](https://github.com/gilt/postcss-less). Although I only tested postcss-scss yet.

webark commented:
@topaxi totally! I was wanting to avoid having to add in that overhead, and with plain CSS, you couldn't do the standard import to include it in the app CSS either. so would have to either tack it only to the app CSS or the vendor CSS manually, or include in a third CSS file with utilizing the headerFor hook. which might not be a bad idea, and could potentially be a better solution for engines long term. Would want others who weigh in who are more familiar and see if it would actually have an impact. Plan CSS is something that should be addressed either right after this PR would get merged, or very soon after.

webark commented:
actually, scratch that. I don't think that it would have any impact on an engine, you'd want to have a single file. And need to have it be imported into an existing app stylesheet to be processed so that you can utilize variables and the likes. So, we should just tac the plain css onto the end of the app css.

webark commented:
@topaxi on further thought. To try to do plain css without a preprocessor, and just doing that ourselves, would be too hacky and would not want to maintain. We will however be basically giving them a preprocessor, which feels weird, but necessary. What do you feel about using cssnext rather then sass or less..?

topaxi commented:
Yes, there's definitely need for a compiler, I'd use postcss (cssnext is only a postcss plugin collection) with as few plugins as possible, and let the addon consumer opt-in for more plugins like cssnext. As I already mentioned, the postcss-selector-namespace plugin I wrote could prefix plain css with the component class, though I'm not sure if we can reuse this 100%. We'd probably also need postcss-import-url to handle the `@import` statements. If we could postcss all cases, and even run postcss after less and sass, that would probably the best way, but I'm not sure if we could do this.

webark commented:
@topaxi Why do you want to run it for every case?

topaxi commented:
@webark If we run postcss in every case, we can support every precompiler for css there is, so we can transform the output of all the precompilers instead of the original files, where we'd need to support less, sass, scss, stylus, etc. But as I said, I'm not sure if we can actually do this (my broccoli/ember-cli knowledge is a bit limited here).

webark commented:
So every preprocsors that uses a unique extension that is not .css, supports nesting with the & selector and uses curly brackets should be supported. This leaves us at the .css files not being supported cause they do not support nesting with the & selector. I had previously assumed this would be only plain CSS files, but that's wrong cause these could be cssnext files also as they use the .css extention, but run a preprocessor on them. We still want the file to be prepped but not processed until after its imported, cause that way you can utilize global variables and helper methods and the likes that are defined in the styles directory in your component CSS files.

I think using that plugin for postcss you wrote to process the .css file would work well here for plain CSS files, and should be lighter weight then brining in sass or something like that. But as long as we keep the same API of the & selector and not let the user customize what that is, I dont see a need to use it for other files. But I totally admit that I could just be missing the point of what your after!!

topaxi commented:
I see your point with `&` being an issue, I always used `:--component` even when using scss.

I think the best way forward and being compatible is only using postcss for plain css for now :)

If you're sure that just wrapping with `.component-class { /* nested styles */ }` is enough I think we're good. (I know there was an issue with this in https://github.com/ebryn/ember-component-css/issues/107 but this might already be fixed?).

topaxi commented:
On another note, it seems your changes don't support sass and stylus indentation style, or am I missing something?

You'd probably have to do something along the lines like here: [lib/broc-component-css-preprocessor.js#L37](https://github.com/ebryn/ember-component-css/blob/master/lib/broc-component-css-preprocessor.js#L37)

webark commented:
I removed sass as an option cause it's kind of deprecated, it would have to use a different output cause it doesn't support curly brackets, and then when importing to the scss file I had, I would have had to import pod-styles.scss rather then just pod-styles, which was annoying.

webark commented:
And I thought stylus supported curly brackets... grr..

topaxi commented:
Stylus supports curly brackets, though I'm not sure if mixing the two styles work.

webark commented:
I'll have to look more into it. I didn't do any testing other then scss files.

webark commented:
You could add a couple of new filters that handle the individual other file extensions. If a type used the same extension but different syntax that would be annoying.

webark commented:
@topaxi what do you feel about allowing the defining of what the self selector is..? I kind of liked your approach of using :--component rather then the &

topaxi commented:
Stylus definitely supports this:

```stylus
body
  background: red;

  .someClass {
    background: blue;
  }
```

Wrapping in curlies doesn't seem to work though:

```stylus
body {
  background: red;

  // Does not work
  .someClass
    background: blue;
}
```

----

I'd definitely prefer `:--component` all the way, it is more explicit than `&`. By allowing `&` we kind of introduce a second meaning to `&`.
What do you think @ebryn?

webark commented:
I think I'm starting to see more of what you are talking about. And now's a good time to do a breaking change if need be. 

So your saying, that we would open up every file, and then loop over each rule definition and either prepend a ".CLASS " or replace :--component with ".CLASS" and that way we don't need to worry about each preprocessors own idiosyncrasies..

webark commented:
Id say the main downside for me is the extra more intensive processing step that requires. But the simplicity and uniform approach to all preprocessors is nice.

webark commented:
if you did a rule that was just an & though, if you pretended that with a ".CLASS " it would still work. Right?

webark commented:
And we couldn't prepend any rules that where nested. Does that plugin you wrote account for that @topaxi?

topaxi commented:
I'm not 100% sure what you mean.

```css
.CLASS & {

}
```

Would result in something like:

```css
.CLASS .my-component {

}
```

What do you mean exactly by "we couldn't prepend any rules that where nested"? Do you have an example?

I agree that the extra processing step may be more CPU intensive, I think we should look at this if we have actual numbers of how much this will slow down the build of a bigger application.

webark commented:
we wouldn't be replaceing the &, and that would be picked up a preprocessor

webark commented:
so if we make a breaking change that a rule that starts with a `&` is no longer supported, and you have to use `:--component` instead, existing usage for people that use a preprocessor will still work cause
```css
& {

}
```
would be changed to
```css
.CLASS & {

}
```
would be picked up by the preprocessor and just turned into 
```css
.CLASS {

}
```

And by nested rules I mean that
```css
:--component {
  color: red;
  .foo {
      color: green;
  }
}
```
should just be processed to be
```css
.CLASS {
  color: red;
  .foo {
    color: green;
  }
}
```
and not
```css
.CLASS {
  color: red;
  .CLASS .foo {
    color: green;
  }
}
```

webark commented:
I think that's what [this](https://github.com/topaxi/postcss-selector-namespace/blob/master/lib/plugin.js#L54-L70) does right @topaxi ?

topaxi commented:
Exactly, this sounds perfect to me. :+1:

webark commented:
ok @topaxi and @ebryn I'll make the change to use postcss-selector-namespace, see if it's still working, do a profile before and after, and update the PR with the new code and my findings.

webark commented:
@topaxi Apparently, at least sass, doesn't recognize and parse a rule that has a class name and a `&`, however, if you keep the selector as `/&|:--component/` it works fine! The performance was not noticiable, and even possibly approved due to less preprocessor time needed to parse all the `&`.

One caveat now is that you have to explicitly declare which file you want to import.. like `@import 'pod-styles.scss'` cause currently it generates a file for every `extensionsForType('css')` extension. so if you leave the extension off of the import, you will get a ```Error: It's not clear which file to import for '@import "pod-styles"'.
Candidates:
  pod-styles.scss
  pod-styles.sass
  pod-styles.css
Please delete or rename all but one of these files.
```
It would be nice to either not create, or delete the file if it is unneeded, or empty. Any suggestions on how to do that?

webark commented:
Ok, I just added a plugin that removes empty files. I don't know if this is the most preferred method for doing so.  @topaxi @ebryn I would like someone with more experience with broccoli to give it a once over however.

webark commented:
so @topaxi and @ebryn all we need now is tacing on the pod-style plain css file. How are we to know to tac it on rather then just let someone import it in the case of cssnext or something? And what's the process for merging this and testing/writing tests/closing out the bugs that this fixes?

choclaudio commented:
bump on the issue. Any ideas on when this might land?

stefanpenner commented:
@webark when you guys are ready for another code-review, ping me.

webark commented:
@stefanpenner I'd say now is a good time. I added this small plugin that removes files when they are empty, and it had a although small, noticeable impact on performance. I wanted to see if there was a better way to do this, or alter the method of including the generated style file, but was wanting @ebryn and @topaxi feedback on that and waiting for them. Could be released as is, and have a change soon to fix, tune that part. But if you @stefanpenner could give it a look over, that would be much appreciated! I implemented your suggestions, and asked a clarifying question for the part you said could come as a later addition.

webark commented:
@ebryn @topaxi @stefanpenner is there anything I can do to get this moving along?

stefanpenner commented:
I should have some time tomorrow afternoon to review and provide feed back. If i forget, feel free to ping me

webark commented:
Sweet thanks!

webark commented:
While I'm waiting for feedback and an architecture decision on the best way to handle including the generated css file for plain css, I'm going to work on filling out the unit tests.

stefanpenner commented on package.json:
```diff
@@ -22,49 +22,41 @@
     "broccoli-asset-rev": "^2.2.0",
-    "ember-cli": "1.13.15",
+    "broccoli-concat": "^2.2.0",
+    "broccoli-funnel": "^1.0.1",
```
many of these dependencies actually seem like prod dependencies, otherwise those installing will not be able to successfully use this addon.

webark replied:
Noticed that today while starting on tests. Will move to dependencies.

stefanpenner commented on index.js:
```diff
@@ -2,147 +2,87 @@
-      inputFiles: this.vendorStaticStyles.concat(['vendor/addons.css']),
-      outputFile: this.options.outputPaths.vendor.css,
-      description: 'Concat: Vendor Styles'
+  treeForAddon(tree) {
```
this is ES6 syntax, and not all nodes supported by ember-cli yet support this syntax.

webark replied:
True. Ok cool.

stefanpenner commented on index.js:
```diff
@@ -2,147 +2,87 @@
+    return new Funnel('app', {
+      srcDir: this._podDirectory(),
+      exclude: ['styles/**/*'],
+      include: this.allowedStyleExtentions.map(function(ext) { return new RegExp(ext + '$'); }),
```
is it possible to use a glob instead?

stefanpenner replied:
https://github.com/ebryn/ember-component-css/pull/123/files#diff-168726dbe96b3ce427e7fedce31bb0bcR31 seems to use globs, we may want to consider consistency.

webark replied:
I'll work on this switch.

stefanpenner commented on index.js:
```diff
@@ -2,147 +2,87 @@
+  _getPodStyleFunnel() {
+    return new Funnel('app', {
+      srcDir: this._podDirectory(),
+      exclude: ['styles/**/*'],
```
node-walk-async doesn't have a fast-path for exclude yet, it shouldn't be a deal breaker, but if someone is motivated Funnel -> walkSync fast-pathing excludes would be really great addition.

stefanpenner commented on index.js:
```diff
@@ -2,147 +2,87 @@
+  _concatenatedPodStyles(wrappedStyles) {
+    var concatenatedStyles = [];
+
+    for (var i = 0; i < this.allowedStyleExtentions.length; i++) {
```
how many extensions are allowed?

webark replied:
three are about 5 css preprocessing extensions? .scss, .sass, .less, . styl, and maybe one more?

stefanpenner commented on index.js:
```diff
@@ -2,147 +2,87 @@
-module.exports = {
-  name: 'ember-component-css',
+    var treeAndNames = new Merge([tree, podNames], {
+      overwrite: true,
```
its quite rare to actually want to override here, nearly always its the sign of a bug or potentially to much work.

webark replied:
overwriting here cause I include a bare file by default that I want to overwrite with the one that is generated. This ensured that the file is always present.

stefanpenner replied:
ok, if there is a explicit reason I'll assume its good :+1:

stefanpenner commented on lib/remove-empty.js:
```diff
@@ -0,0 +1,27 @@
+}
+
+RemoveFiles.prototype.build = function() {
+  for (var i = 0; i < this.inputPaths.length; i++) {
```
same as above, persistent filter variant is likely the best bet. This one is likely quite easy to implement that way https://github.com/stefanpenner/broccoli-persistent-filter/blob/master/index.js#L62-L91 should be a reasonable good guide.

The idea is to only apply the require patch, to the last output dir. Rather then having to reconstruct the whole thing.

stefanpenner replied:
Again sorry there is not yet a functioning nicer high level API for this, hopefully after my next few speaking engagements I will again have free time to finish up that effort. I feel the current state is extremely crappy and largely because I haven't had a chance to finish my effort yet...

webark commented:
@stefanpenner thanks for reviewing this! I'll start working on those changes tomorrow while I'm filling out some tests. Thanks again!

webark commented:
@stefanpenner I moved the generating of the pod-names.js file that gets imported into the initializer to use persistentOutput and fs-tree-diff. Let me know if that's what you where thinking for that one. And I'll either find a better solution for the removing of the files empty files with the other one, or switch it to use the same methodology.

rwjblue commented on addon/initializers/component-styles.js:
```diff
@@ -0,0 +1,36 @@
+
+  ComponentLookup.reopen({
+    componentFor(name, owner) {
+      if (podNames[name] && !owner.application.hasRegistration('component:' + name)) {
```
On Ember 2.3 and higher, owner.application should not be used. You should use owner.hasRegistration and owner.register. To enable usage of the same API across older Ember versions you can use the ember-getowner-polyfill.

rwjblue replied:
Also, wouldn't it be easier to call super and check if the return value is undefined instead of manually checking registration like this? Something like:

    let FoundComponent = this._super(...arguments);
    If (podNames[name] && !FoundComponent) { 
      FoundComponent = Component;
      owner.register('component:' + name, FoundComponent);
    }
    
    return FoundComponent;

webark replied:
I had tried this it wasn't working for me. I think cause _lookupFactory does more then just return the constructor. I would have to call super again at the end, which calling it twice felt weird.

rwjblue commented on addon/initializers/component-styles.js:
```diff
@@ -0,0 +1,36 @@
+  ComponentLookup,
+} = Ember;
+
+export function initialize() {
```
This work should not be done inside the initialize function (if it is it will be reran once for every test. It should be done in the root of the module so that it is only done once when the module is evaluated (and before the application itself is created).

rwjblue commented on index.js:
```diff
@@ -2,147 +2,87 @@
-    options.registry = this.registry;
-    var preprocessedStyles = preprocessCss(stylesAndVendor, '/app/styles', '/assets', options);
+  included(app) {
+    this._super.included(app);
```
Should be:

    this._super.included.apply(this, arguments);

rwjblue commented on index.js:
```diff
@@ -2,147 +2,87 @@
-    if (!podModulePrefix) { return 'app'; }
-    var podPath = podModulePrefix.substr(podModulePrefix.indexOf('/') + 1);
-    return path.join('app', podPath);
+    return this._super.treeForAddon.call(this, treeAndNames);
```
Should use apply and pass all args through.

webark replied:
I was following the instructions for these hook types [here](https://github.com/ember-cli/ember-cli/blob/master/ADDON_HOOKS.md#treefor-cont)

The passed in tree need to get modified to include other trees, and redefining the passed in argument calling super with arguments wasn't working for me.

rwjblue commented on index.js:
```diff
@@ -2,147 +2,87 @@
-      styles: '',
-      extension: ''
-    };
+    return this._super.treeForStyles.call(this, removeEmptyFiles);
```
^^ apply

webark replied:
^^

rwjblue commented on lib/component-names.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  path(actualPath) {
```
Es6 syntax (wont work on node 0.10)

rwjblue commented:
This is looking great! I'd love to see this land on master and released (with a major version bump) as a prerelease so folks can start test/using it (and we can track issues/PR's independently).

webark commented:
@rwjblue and @stefanpenner all of you suggestions have been addressed I believe. @ebryn or @topaxi what do you feel about the suggestion to prerelease this with a major version bump? I think it's read for that. I'm going to be replacing the tests over the next couple of days, and an architecture decision needs to be made about how to support plain css files (if at all) when no preprocessor is included. But it would be nice to get this into master sooner rather then later.

ebryn commented:
I've created a `stable` branch for the existing 0.1.x code, rebased and merged this PR to master

ebryn commented:
We've got a beta release for this work as well... `ember-component-css@0.2.0-beta.1`
